### PR TITLE
[NEAT-41] Export to CDF and Integration Test

### PR DESCRIPTION
**Pull request by doctrino**
Overall got the DMS Export to DMS to work in the first iteration (still more validation to add). On the way I ran into multiple obstacles that resulted in the following. 

* Added check to InformationRules that all classes have at least one property.
* Added a few properties to the InformationRules sheet for the classes with no properties.
* Implement `DMSExporters.export_to_cdf`.
* Added validation to ensure direct relations are nullable, updated conversion from Information Architect to reflect this. (Even if the Information Model requires a direct relation, it is set to nullable).

### CODE REVIEW

File: cognite/neat/rules/exporters/_loaders.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    VersionedDataModelingId,
+    ViewId,
+)
+from cognite.client.utils.useful_types import SequenceNotStr
+
+T_ID = TypeVar("T_ID", bound=str | (int | (DataModelingId | (InstanceId | VersionedDataModelingId))))
+T_WritableCogniteResourceList = TypeVar("T_WritableCogniteResourceList", bound=WriteableCogniteResourceList)
+
+
+class ResourceLoader(
```
Comment (doctrino) [inline cognite/neat/rules/exporters/_loaders.py:46]: This is steeling a concept from `cdf-toolkit`. However, I thing it is equally useful for us to be able to tell the user results of dry runs and have a standardized interface towards cognite resources.

File: cognite/neat/rules/exceptions.py [hunk 2]
```diff
@@ -2022,6 +2022,24 @@ def __init__(self, missing_fields: set[str], verbose: bool = False):
         super().__init__(self.message)
 
 
+class ClassNoProperties(NeatException):
```
Comment (nikokaoja) [inline cognite/neat/rules/exceptions.py:2025]: In principle you might have class without a property if that class has a parent. Example GeneratingUnit -> WindGeneratingUnit(GeneratingUnit), so basically either Class must have:
- either properties
- or parent(s)
- or both parent(s) and properties

File: cognite/neat/rules/exceptions.py [hunk 3]
```diff
@@ -2022,6 +2022,24 @@ def __init__(self, missing_fields: set[str], verbose: bool = False):
...
+    """ """
+
+    type_: str = "ClassNoProperties"
```
Comment (nikokaoja) [inline cognite/neat/rules/exceptions.py:2028]: So you would rename this into `ClassNoPropertiesNoParents`

File: cognite/neat/rules/exporters/_data_classes.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+from abc import ABC
+from dataclasses import dataclass, field
+from functools import total_ordering
+
+
```
Comment (nikokaoja) [inline cognite/neat/rules/exporters/_data_classes.py:5]: should we call this module `_models` ? not really religious

File: cognite/neat/rules/exporters/_loaders.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+
+class SpaceLoader(DataModelingLoader[str, SpaceApply, Space, SpaceApplyList, SpaceList]):
+    resource_name = "spaces"
+
+    @classmethod
+    def get_id(cls, item: Space | SpaceApply) -> str:
+        return item.space
+
+    def create(self, items: Sequence[SpaceApply]) -> SpaceList:
```
Comment (nikokaoja) [inline cognite/neat/rules/exporters/_loaders.py:103]: nicely org CRUD operations! Like it

File: cognite/neat/rules/exporters/_loaders.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        return self.client.data_modeling.containers.retrieve(cast(Sequence, ids))
+
+    def update(self, items: Sequence[ContainerApply]) -> ContainerList:
+        return self.create(items)
+
+    def delete(self, ids: SequenceNotStr[ContainerId]) -> list[ContainerId]:
+        return self.client.data_modeling.containers.delete(cast(Sequence, ids))
+
+
+class DataModelLoader(DataModelingLoader[DataModelId, DataModelApply, DataModel, DataModelApplyList, DataModelList]):
```
Comment (nikokaoja) [inline cognite/neat/rules/exporters/_loaders.py:229]: Create `cognite/neat/rules/exporters/_dms/` and bundle dms specific things together. We use this approach for owl importer. Keeps things nice and clean

File: cognite/neat/rules/exporters/_loaders.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    def retrieve(self, ids: SequenceNotStr[DataModelId]) -> DataModelList:
+        return self.client.data_modeling.data_models.retrieve(cast(Sequence, ids))
+
+    def update(self, items: Sequence[DataModelApply]) -> DataModelList:
+        return self.create(items)
+
+    def delete(self, ids: SequenceNotStr[DataModelId]) -> list[DataModelId]:
+        return self.client.data_modeling.data_models.delete(cast(Sequence, ids))
+
+    def are_equal(self, local: DataModelApply, remote: DataModel) -> bool:
```
Comment (nikokaoja) [inline cognite/neat/rules/exporters/_loaders.py:248]: I like this `local`  and `remote`. We should remebre to use same naming for Asset Loader.

File: cognite/neat/rules/exporters/_rules2dms.py [hunk 8]
```diff
@@ -38,4 +51,64 @@ def export_to_file(self, filepath: Path) -> None:
                 zip_ref.writestr(f"data_models/{container.external_id}.container.yaml", container.dump_yaml())
 
     def export(self) -> DMSSchema:
-        return self.rules.as_schema()
+        if self._schema is None:
+            self._schema = self.rules.as_schema()
+        return self._schema
+
+    def export_to_cdf(self, client: CogniteClient, dry_run: bool = False) -> Iterable[UploadResult]:
```
Comment (nikokaoja) [inline cognite/neat/rules/exporters/_rules2dms.py:58]: Nice

File: cognite/neat/rules/models/_rules/information_rules.py [hunk 9]
```diff
@@ -245,6 +245,15 @@ def validate_schema_completeness(self) -> Self:
 
         return self
 
+    @model_validator(mode="after")
+    def validate_class_has_properties(self) -> Self:
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/information_rules.py:249]: `has_properties_or_parent`, ref the first comment

**Review by nikokaoja — APPROVED** [review]

File: cognite/neat/rules/exporters/_loaders.py (hunk 7 above)
Comment (doctrino) [reply cognite/neat/rules/exporters/_loaders.py]: That is already there in the new AssetLoader implementation.

**Pull request merged**